ferrocene/ci/scripts/detect-unused-cargo-patches.py

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+
+import subprocess
+import tomllib
+from collections import defaultdict
+
+
+def find_files(name):
+    cmd = subprocess.run(
+        ["git", "ls-tree", "HEAD", "--name-only", "-r", "-z"],
+        stdout=subprocess.PIPE,
+        check=True,
+        text=True,
+    )
+
+    # The -z flag changes the output to separate file names with a null byte, preventing files with
+    # spaces or newlines from messing things up.
+    for file in cmd.stdout.split("\0"):
+        if file == name or file.endswith(f"/{name}"):
+            yield file
+
+
+def find_unused_patches(file):
+    with open(file, "rb") as f:
+        try:
+            toml = tomllib.load(f)
+        # Probably due to conflict markers:
+        except tomllib.TOMLDecodeError:
+            print(f"note: skipping {file} as it is not valid TOML")
+            return
+
+    try:
+        unused_patches = toml["patch"]["unused"]
+    except KeyError:
+        return
+
+    package_versions = defaultdict(list)
```

Comment:
I am somewhat alarmed by this line but it does seem to be valid python, somehow.

Reply:
When you try to access a missing element, `defaultdict` creates the new element with the function provided to it. In this case the function is `list`, creating an empty list.

It's equivalent to this code, should we switch to it?

```suggestion
    package_versions = defaultdict(lambda: [])
```

Reply:
I think it's fine, Python just unsettles me sometimes. 😅